### PR TITLE
fix(tianmu): fix group by lead to wrong result (#1107)

### LINKED ISSUE #1107: bug: Group by lead to wrong result
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```sql
create table A(id int,UnitSN varchar(10),CTIME datetime);
insert into A values(1,'A1','2022-12-09 01:00:00');
insert into A values(2,'A2','2022-12-09 02:00:00');
insert into A values(3,'A3','2022-12-09 03:00:00');
insert into A values(4,'A4','2022-12-09 04:00:00');
insert into A values(5,'A5','2022-12-09 05:00:00');

create table B(id int,AID int,Name varchar(10),value varchar(10));
insert into B values(1,1,'V1','XXXXXXX1');
insert into B values(2,1,'V2','XXXXXXX2');
insert into B values(3,1,'V3','XXXXXXX3');
insert into B values(4,1,'V4','XXXXXXX4');
insert into B values(5,1,'V5','XXXXXXX5');

mysql> select a.ID,
    ->        a.UNITSN,
    ->        case upper(b.NAME) when 'V1' then b.VALUE else null end as V1,
    ->        case upper(b.NAME) when 'V2' then b.VALUE else null end as V2,
    ->        case upper(b.NAME) when 'V3' then b.VALUE else null end as V3,
    ->        case upper(b.NAME) when 'V4' then b.VALUE else null end as V4,
    ->        case upper(b.NAME) when 'V5' then b.VALUE else null end as V5
    ->   from A a
    ->   left join B b
    ->     on a.ID = b.AID
    ->  group by a.ID, a.UNITSN;
+------+--------+----------+----------+----------+----------+----------+
| ID   | UNITSN | V1       | V2       | V3       | V4       | V5       |
+------+--------+----------+----------+----------+----------+----------+
|    1 | A1     | XXXXXXX1 | XXXXXXX2 | XXXXXXX3 | XXXXXXX4 | XXXXXXX5 |
|    4 | A4     | NULL     | NULL     | NULL     | NULL     | NULL     |
|    2 | A2     | NULL     | NULL     | NULL     | NULL     | NULL     |
|    3 | A3     | NULL     | NULL     | NULL     | NULL     | NULL     |
|    5 | A5     | NULL     | NULL     | NULL     | NULL     | NULL     |
+------+--------+----------+----------+----------+----------+----------+
5 rows in set (0.00 sec)
```

### Expected behavior

```sql
mysql> select a.ID,
    ->        a.UNITSN,
    ->        case upper(b.NAME) when 'V1' then b.VALUE else null end as V1,
    ->        case upper(b.NAME) when 'V2' then b.VALUE else null end as V2,
    ->        case upper(b.NAME) when 'V3' then b.VALUE else null end as V3,
    ->        case upper(b.NAME) when 'V4' then b.VALUE else null end as V4,
    ->        case upper(b.NAME) when 'V5' then b.VALUE else null end as V5
    ->   from A a
    ->   left join B b
    ->     on a.ID = b.AID
    ->  group by a.ID, a.UNITSN;
+------+--------+----------+------+------+------+------+
| ID   | UNITSN | V1       | V2   | V3   | V4   | V5   |
+------+--------+----------+------+------+------+------+
|    1 | A1     | XXXXXXX1 | NULL | NULL | NULL | NULL |
|    2 | A2     | NULL     | NULL | NULL | NULL | NULL |
|    3 | A3     | NULL     | NULL | NULL | NULL | NULL |
|    4 | A4     | NULL     | NULL | NULL | NULL | NULL |
|    5 | A5     | NULL     | NULL | NULL | NULL | NULL |
+------+--------+----------+------+------+------+------+
5 rows in set (0.00 sec)
```

### How To Reproduce

_No response_

### Environment

./mysqld  Ver 5.7.36-StoneDB for Linux on x86_64 (build-)
build information as follow:
        Repository address: https://github.com/stoneatom/stonedb.git:HEAD
        Branch name: HEAD
        Last commit ID: 2bb8c6023
        Last commit time: Date:   Mon Dec 5 22:16:35 2022 +0800
        Build time: Date: Fri Dec  9 07:12:05 UTC 2022

### Are you interested in submitting a PR to solve the problem?

- [X] Yes, I will!

### PR DESCRIPTION
**Pull request by adofsauron**
## Summary about this PR

Issue Number: close #1107 

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

File: storage/tianmu/core/aggregator_basic.cpp [hunk 1]
```diff
@@ -543,8 +543,13 @@ void AggregatorListT::PutAggregatedValue(unsigned char *buf, const types::BStrin
     *((unsigned short *)buf) = v.len_;
     if (v.len_ > 0)
       std::memcpy(buf + 2, v.val_, v.len_);
-    else
-      buf[2] = 1;  // empty string indicator (non-null)
+    else {
+      if (v.IsNull() && IgnoreNulls()) {
```
Comment (RingsC) [inline storage/tianmu/core/aggregator_basic.cpp:547]: Add some explaination on these const value, 1, 2 and so on.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/aggregator_basic.cpp (hunk 1 above)
Comment (RingsC) [reply storage/tianmu/core/aggregator_basic.cpp:547]: 2 changes to sizeof (unsigned short)

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/aggregator_basic.cpp (hunk 1 above)
Comment (RingsC) [reply storage/tianmu/core/aggregator_basic.cpp:547]: And pls, refactor type-casting in this file by using c++ style type-cast. Thanks.

**Review by RingsC — CHANGES_REQUESTED** [review]

**Review by RingsC — APPROVED** [review]
LGTM

**Review by lujiashun — APPROVED** [review]
LGTM

**Pull request merged**